lib/ansible/modules/files/file.py

```diff
@@ -319,35 +319,44 @@ def get_state(path):
-                            changed |= recursive_set_attributes(b_fsname, follow, file_args, mtime, atime)
-
-                        # Change perms on the file pointed to by the link
-                        tmp_file_args = file_args.copy()
-                        tmp_file_args['path'] = to_native(b_fsname, errors='surrogate_or_strict')
-                        changed |= module.set_fs_attributes_if_different(tmp_file_args, changed, expand=False)
-                        changed |= update_timestamp_for_file(tmp_file_args['path'], mtime, atime)
+    try:
+        for b_root, b_dirs, b_files in os.walk(b_path):
+            for b_fsobj in b_dirs + b_files:
+                b_fsname = os.path.join(b_root, b_fsobj)
+                if not os.path.islink(b_fsname):
+                    tmp_file_args = file_args.copy()
+                    tmp_file_args['path'] = to_native(b_fsname, errors='surrogate_or_strict')
+                    changed |= module.set_fs_attributes_if_different(tmp_file_args, changed, expand=False)
+                    changed |= update_timestamp_for_file(tmp_file_args['path'], mtime, atime)
+
+                else:
+                    # Change perms on the link
+                    tmp_file_args = file_args.copy()
+                    tmp_file_args['path'] = to_native(b_fsname, errors='surrogate_or_strict')
+                    changed |= module.set_fs_attributes_if_different(tmp_file_args, changed, expand=False)
+                    changed |= update_timestamp_for_file(tmp_file_args['path'], mtime, atime)
+
+                    if follow:
+                        b_fsname = os.path.join(b_root, os.readlink(b_fsname))
+                        # The link target could be nonexistent
+                        if os.path.exists(b_fsname):
+                            if os.path.isdir(b_fsname):
+                                # Link is a directory so change perms on the directory's contents
+                                changed |= recursive_set_attributes(b_fsname, follow, file_args, mtime, atime)
+
+                            # Change perms on the file pointed to by the link
+                            tmp_file_args = file_args.copy()
+                            tmp_file_args['path'] = to_native(b_fsname, errors='surrogate_or_strict')
+                            changed |= module.set_fs_attributes_if_different(tmp_file_args, changed, expand=False)
+                            changed |= update_timestamp_for_file(tmp_file_args['path'], mtime, atime)
+    except RuntimeError as e:
+        # on Python3 "RecursionError" is raised which is derived from "RuntimeError"
+        # TODO once this function is moved into the common file utilities, this should probably raise more general exception
+        raise AnsibleModuleError(
+            results={'msg': "An error occured when trying recursively setting attributes on %s. Original error was: '%s'" % (to_native(b_path), to_native(e))}
```

Comment:
you should not need 'An error occurred' we already prefix these with ERROR or FAILURE

shouldn't we fail_json at this point?

Reply:
> you should not need 'An error occurred' we already prefix these with ERROR or FAILURE

With this PR I get:
```
fatal: [localhost]: FAILED! => {"changed": false, "msg": "An error occured when trying recursively setting attributes on /tmp/demo-recurse. Original error was: 'maximum recursion depth exceeded while calling a Python object'"}
```

How would you re-phrase?

> 
> shouldn't we fail_json at this point?

Raising `AnsibleModuleError` triggers `fail_json`, see https://github.com/ansible/ansible/blob/f0c59d4cc08185ecaa6d1cf71f47c2442597bb7b/lib/ansible/modules/files/file.py#L238

Reply:
`Could not recursively set attri...`